### DISCUSSION TITLE
Emulating a Proprietary USB Game Input Device

### DISCUSSION THREAD

**Discussion by 641i130** [hathach/tinyusb#1508] [Q&A]
I'll try to keep this as simplified and as to the point as possible.

I'm attempting to emulate a rare Japanese arcade machine USB IO card. The way I'd like this to be configured (through the TinyUSB Library) is like this: https://clbin.com/OS0td
TL;DR
My goal is to make a 255 Vendor Specific Class USB RP2040-based device that has 2 endpoints shown below:
```
Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
```
I'm currently able to get the device to show up with the right vendor/product id, but using the webusb serial example, I found that it wasn't allowing me to use the endpoint it made. It was 'in use'. 

How do I get a better understanding of creating endpoints? How should I develop endpoints based on the given tinyusb examples for my specific use case? Which examples would help me? This device does not use HID, so I'm unsure what's the closest...

**Comment by pigrew** [comment]
The `vendor_device` class might be usable as-is, or perhaps you can copy and modify it for your use.

Have you fully reverse engineered the protocol? That would be the first task... It sounds like you likely cannot install extra software on your arcade machine itself, so you'd need a stand-alone logic analyzer or USB analyzer to capture example data?

**Reply by 641i130** [reply]
I have actually! I made a tool that shows the buffer values. I just want to get to the point where I can see the endpoint values populate my tool. The protocol is documented here with the tool I made:
https://github.com/641i130/ttx4-usbio

**Comment by zlittell** [comment]
I think using the simple in/out device hid example would be a better starting point. Theres no extra fluff. You can use the webusb example as a reference to "Vendor class" specific things (ignoring all CDC stuff)

You should be able to hardcode a lot of the descriptors instead of using the helpful predefined functions (and probably should since you are emulating another device). Then you will just handle the incoming packets in the application code.

**Reply by 641i130** [reply]
Ok, I'll attempt this! Thank you!

**Reply by 641i130** [reply]
I'm getting errors with the HID functions... I'm not sure what to replace the endpoint making HID functions with... I'll need to dive much deeper than I thought...

**Reply by pigrew** [reply]
One big caveat about HID is that it uses control and interrupt endpoints, whereas you (AFAIK) are using bulk. This means you need to change the EP types in the descriptors.... The endpoint definitions are pre-processor macros that statically create the descriptor table at compile-time.

**Reply by 641i130** [reply]
Sorry about the late response (just now getting back into this project),

so does bulk have a different CLASS configuration? If I'm just doing a VENDOR specific class like so: 
![image](https://user-images.githubusercontent.com/46945263/189512962-71dccf1c-a52a-4076-a15e-aa84713417fa.png)
Do I need any more parts to make sure I'm getting the data to match the lsusb output?
